### PR TITLE
Add FPGA algorithm

### PR DESCRIPTION
**Pull request by EExiaLee**
1 [Feature] Add algorithm(paillier_clustar_fpga) which uses FPGA to speed up Paillier 
2 [Optimize] Optimize heu::lib::numpy::Evaluator::MatMul in Vectorized SPI case 
3 [Feature] New apis are added in all algorithms to make them still work in 2 
4 [Add] Add test case Mul in NpBenchmarks

### CODE REVIEW

**Comment by 6fj** [comment]
recheck

File: heu/library/numpy/evaluator.cc [hunk 1]
```diff
@@ -242,11 +245,22 @@ auto DoCallMatMul(const CLAZZ& sub_evaluator, const M1& mx, const M2& my,
...
+          std::vector<decltype(&res[0])> sum_vec;
+          sum_vec.reserve(res.size());
+          for (size_t j = 0; j < res.size(); ++j) {
+            sum_vec.push_back(&res[j]);
+          }
 
-        for (size_t j = 1; j < res.size(); ++j) {
-          // todo: use binary reduce
-          sub_evaluator.AddInplace({sum}, {&res[j]});
+          sub_evaluator.CalcSum(sum, sum_vec);
```
[the remark this answers is not in the document]
Comment (EExiaLee) [reply heu/library/numpy/evaluator.cc:261]: 好的

File: heu/library/numpy/evaluator.cc [hunk 2]
```diff
@@ -242,11 +245,22 @@ auto DoCallMatMul(const CLAZZ& sub_evaluator, const M1& mx, const M2& my,
 
         auto res = sub_evaluator.Mul(in_x[row], in_y[col]);
         auto sum = &res[0];
+        //
+        if (sub_evaluator.GetSchema() == "ipcl") {
```
[the remark this answers is not in the document]
Comment (EExiaLee) [reply heu/library/numpy/evaluator.cc:249]: 此处的设想是ipcl也实现CalcSum, 然后这里用一致的实现，不进行schema的判定了。

[the remark this answers is not in the document]
Comment (usafchn) [reply heu/library/numpy/evaluator.cc:249]: ok

**Review by usafchn — APPROVED** [review]

**Pull request merged**